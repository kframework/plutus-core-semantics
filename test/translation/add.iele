contract Add {

  define @dispatch(%closure, %param) {
    %fptr = load %closure

    %cond = cmp eq %fptr, 1
    br %cond, label1

    %cond = cmp eq %fptr, 2
    br %cond, label2

    label2:
      call @fresh2(%closure, %param)

    label1:
      call @fresh1(%closure, %param)
  }

  define @init() {
    %alloc_counter_ptr = 0
    store 1, %alloc_counter_ptr
  }

  define @alloc() {
    %alloc_counter_ptr = 0
    %alloced = load %alloc_counter_ptr
    %ret = %alloced
    %alloced = add %size, %alloced
    store %alloced, %alloc_counter_ptr
    ret %ret
  }

  define @add() {
    %ret = call @alloc(1)
    store 1, %ret
    ret %ret
  }

  define @fresh1(%x) {
    %ret = call @alloc(2)
    store 2, %ret
    %fresh3 = add %ret, 1
    store %x, %fresh3
    ret %ret
  }

  define @fresh2(%env, %y) {
    %addr = add %env, 0
    %fresh4 = load %addr
    %ret = add %fresh4, %y
    ret %ret
  }

}