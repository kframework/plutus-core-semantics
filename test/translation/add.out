<T>
  <k> ( define add ( lam x  .VarList ( lam y  .VarList ( builtin addInteger x y ) ) ) ) </k>
</T>
